### PR TITLE
Recursive clean-up on uninstalls

### LINKED ISSUE #745: Uninstall to clear parent folder when empty
From https://github.com/jspm/github/issues/48:

> If you uninstall a GitHub repo named `x/y`, then the `y` folder is removed, but not the empty `x` folder.

### PR DESCRIPTION
**Pull request by crisptrutski**
Closes #745

### CODE REVIEW

File: lib/install.js [hunk 1]
```diff
@@ -869,6 +869,34 @@ function showVersions(forks) {
 }
 exports.showVersions = showVersions;
 
+/* Recursively remove directory, all those above it, if they are empty.
+ * Takes optional `stopDir` to terminate at. */
+function cascadeDelete(dir, stopDir) {
```
Comment (guybedford) [inline lib/install.js]: Perhaps let's put this into the common.js file?

**Comment by guybedford** [comment]
:+1:

**Comment by crisptrutski** [comment]
@guybedford managed to find a subtle issue where parallel deletes occur in the gap between `exists` check and `rmdir`. 

Might not even be worth calling `exists` anyway now, or `readdir`, and just use the error codes. Less code

**Comment by guybedford** [comment]
Well caught, yeah async needs a lot of thought. Error codes via a single access is the recommended approach for this stuff anyway. Do put some thought to symlinks as well here - they shouldn't be followed.

**Comment by crisptrutski** [comment]
Not sure we should worry about symlinks here - if symlinking to a shared deps folder, think that removing empty directories is still OK. 

We could check in advance if any of the parent directories are symlinks, and then skip all of them. Is that what you mean? Can't think of anything else that makes sense to me

**Pull request merged**